### ISSUE TITLE
_PKG_CONFIG_LIBS might be empty

### ISSUE THREAD

**Issue by nim65s** [jrl-umi3218/jrl-cmakemodules#350]
Hi,

This line assumes that `_PKG_CONFIG_LIBS` is not empty:

https://github.com/jrl-umi3218/jrl-cmakemodules/blob/a5c65a0ee51191be3f2e2667b95830706c211bb2/package-config.cmake#L136

So, in the case of example-robot-data, this fails with:
```
CMake Error at cmake/package-config.cmake:136 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  cmake/base.cmake:155 (SETUP_PROJECT_PACKAGE_FINALIZE)
  CMakeLists.txt:9223372036854775807 (SETUP_PROJECT_FINALIZE_HOOK)
```

(nice line number, btw)

**Comment by wxmerkt** [comment]
I am just now opening a pull request for this...

**Comment by jcarpent** [comment]
I think #352 is more suitable. @nim65s I've checked on example-robot-data and it works fine.

**Comment by wxmerkt** [comment]
#352 will not work to solve the issue we are seeing on the buildservers.

**Comment by jcarpent** [comment]
What did you change so far? I don't get the point here sorry.

**Comment by wxmerkt** [comment]
The `SETUP_.._FINALIZE` macro, where the code currently is, gets called at the end of the configuration. I.e., before the library (libxxx.so) has been built. As such, the `find_library` statement will not find the absolute path (cf. e.g. https://github.com/ros/rosdistro/pull/24287#issuecomment-607300670). 

When we run a local compile, we may still have copies of the library _somewhere_, so it may appear to work. The fix in #351 moves it to the Config. I tried several variants and this was the one that worked reliable. If you have another suggestion, happy to try and test. However, this regression in the projectConfig have now taken two days of debugging this week so I am being time constrained.

**Comment by gergondet** [comment]
I think the issue @wxmerkt is seeing happens for libraries that are built by the project in question (in that case eigenpy). In such cases the correct fix would be to use a generator expression to get the absolute installation path of the library rather than trying to use `find_library` (there's a lot of way this could report the wrong library whereas the generator expression will get it right)

@wxmerkt out of curiosity, why not just change the consumer of eigenpy so that it uses CMake and not pkg-config?

**Comment by wxmerkt** [comment]
I think the generator expression could work. If someone can provide an alternate suggestion/fix to #351, I am happy to test.

(We cannot change consumers (in this case MoveIt) due to a regression in the packaging. The use case in question was that for EigenPy a CMake config existed, which got replaced by the new CMake exports which differed in exporting the _LIBRARiES variable. I agree though that in future consumers should move to the CMake interfaces)

**Comment by jcarpent** [comment]
@gergondet Thanks for the comments. How will you get in an automatic way this generated variable?
By the way, @wxmerkt does this exported variable should also contained the dependencies?

**Comment by wxmerkt** [comment]
> does this exported variable should also contained the dependencies?

Yes, that is how the `${project_LIBRARIES}` usually behave. E.g. if I link against `pinocchio_LIBRARIES`, I expect it to have urdfdom, eigenpy, boost_python in that variable (so that I dont manually have to think about where they are...)

**Comment by gergondet** [comment]
@jcarpent sorry I though there was a straight-forward [property](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#target-dependent-queries) that could be queried for that but it seems it might be a little more involved (it's quite late on that side of the world so I didn't look deeply into it)

@wxmerkt if you are just looking for a quick fix you can "reset"  `eigenpy_LIBRARIES` using `PROJECT_EXTRA_MACROS` within eigenpy itself to something more sensible

**Comment by jcarpent** [comment]
@wxmerkt I guess you just need the content of `pkg-config --libs eigenpy `?

**Comment by wxmerkt** [comment]
>  I guess you just need the content of pkg-config --libs eigenpy ?

No - CMake returns absolute paths while `pkg-config --libs eigenpy` would return `-leigenpy`. This was the reason for the fix: https://github.com/stack-of-tasks/eigenpy/commit/65ade259193c85c79701530bae70f80bc3af05d8

We need a fix that does the same as #351. Best way to debug this is in a clean Docker - the created `eigenpyConfig.cmake` needs to correctly have the full install paths for the libraries.

**Comment by jcarpent** [comment]
@wxmerkt `pkg-config --libs eigenpy` provides this output:
`-L/local/install/lib -leigenpy -Wl,-undefined,dynamic_lookup,/usr/local/lib/libboost_python27-mt.dylib`

**Comment by jcarpent** [comment]
Because in the eigenpy.pc, we have:
`Libs: -L${libdir}  -leigenpy -Wl,-undefined,dynamic_lookup,/usr/local/lib/libboost_python27-mt.dylib`

**Comment by wxmerkt** [comment]
> @wxmerkt pkg-config --libs eigenpy provides this output:
-L/local/install/lib -leigenpy -Wl,-undefined,dynamic_lookup,/usr/local/lib/libboost_python27-mt.dylib

Yes, and this won't work with CMake unless the full path (without considering the library dirs with capital L) is specified for each lib. Which is the reason for the fix in https://github.com/stack-of-tasks/eigenpy/commit/65ade259193c85c79701530bae70f80bc3af05d8. A similar fix to avoid the regression is in #350.
